### ISSUE TITLE
Very slow compression speed for level 14+ on some specific datasets

### ISSUE THREAD

**Issue by KrzysFR** [Cyan4973/zstd#130]
I'm currently doing some tests and benchmarks on the best way to compress vectors of numerical data using zstd (and various filters to help compression, such as delta or shuffle).

I found a few oddities while trying various sample sets, and I'm not sure if this is expected or not.

_note: All the tests below are done with the 0.5.1 release._

The most obvious anomaly I found was when compressing vectors of 64-bit floats produced by going from `100.0` and randomly removing a tiny amount at each step (a few tens of %) while keeping the full precision (ie: 17 decimals), i.e: the sample set looks like this (whith each value encoded as an IEEE 64-bit decimal:

```
XS: [
  99.9996492024556, 99.9996492024556, 99.9996492024556, 99.9996492024556, 99.9996492024556,
  99.9959685493382, 99.9934385250828, 99.9915028228664, 99.9913987684419, 99.9876946097741,
  99.9832594119447, 99.9832594119447, 99.9827409006435, 99.9827409006435, 99.9792033732376,
  99.9792033732376, 99.9792033732376, 99.9792033732376, 99.9792033732376, 99.9779770870381,
  ...,
  63.9185913211865, 63.9185913211865, 63.9185913211865, 63.9183349283913, 63.9173471531838,
  63.9129878772782, 63.9129878772782, 63.9129878772782, 63.9129878772782, 63.910876784327,
  63.910876784327, 63.9107775813923, 63.9107775813923, 63.9107775813923, 63.9107775813923,
  63.9107775813923, 63.9077360819731, 63.9077360819731, 63.9059456902976, 63.9059456902976
]
```

I'm also compressing the delta-encoded vector (0 = no change):
```
DELTA(XS): [
  99.9996492024556, 0, 0, 0, 0,
  -0.00368065311744203, -0.00253002425540672, -0.00193570221631489, -0.000104054424497235, -0.00370415866780149,
  -0.00443519782946566, 0, -0.000518511301152103, 0, -0.00353752740591062,
  0, 0, 0, 0, -0.00122628619951115,
  ...,
  -0.000363898205272051, 0, 0, -0.000256392795243698, -0.000987775207491381,
  -0.00435927590558549, 0, 0, 0, -0.00211109295120337,
  0, -9.92029346988943E-05, 0, 0, 0,
  0, -0.00304149941915455, 0, -0.00179039167556283, 0
]
```

Now my benchmark compresses both vectors (N=28,800) encoded as 4-byte or 8-byte elements (115KB / 230KB), using multiple codecs (lz4, zstd, zlib) and also filtering (none, blosc-like shuffle), and measure both the ratio and time to encode.

_The full benchmark results can be found here: https://gist.github.com/KrzysFR/0f6835c7a8d0f19dbdc3  (warning: lots of data and ASCII charts!)_

The combination of delta-encoding + shuffling on 64-bit floats with full precision induce a very visible slowdown for levels 14 up to 21, going from 15ms at level 13 to 280ms at level 14 (up to 21). This bump is still there but less visible when rounding all the numbers to keep only 3 digits.

Below are some charts that show the results.

**Comparison of compression time**
_top: original data set with full precision, bottom: rounded to 3 decimals_
![image](https://cloud.githubusercontent.com/assets/1662235/13220581/432c7d76-d977-11e5-97df-0e9d78d03b1e.png)
![image](https://cloud.githubusercontent.com/assets/1662235/13220584/4974f956-d977-11e5-8523-dc92dc56a838.png)

The yellow line is clearly causing troubles for levels `14` and up. We can also see that ZLib has some issues with it from level `7` and up

Here it is again, but using a log scale for the time:
![image](https://cloud.githubusercontent.com/assets/1662235/13220944/1f430ce8-d979-11e5-9459-1c546a75039e.png)

**Comparison of ratios**
_top: original data set with full precision, bottom: rounded to 3 decimals_
![image](https://cloud.githubusercontent.com/assets/1662235/13220559/1f878906-d977-11e5-9c82-b2339b41070d.png)
![image](https://cloud.githubusercontent.com/assets/1662235/13220567/288497ce-d977-11e5-819f-8b338c78765e.png)

**Full results**
![image](https://cloud.githubusercontent.com/assets/1662235/13220613/63800d36-d977-11e5-956c-c41e332a8e79.png)

**Comment by KrzysFR** [comment]
You can find the sample that corresponds to the yellow line in this gist: https://gist.github.com/KrzysFR/bb7ffe03bdffbd023422 (base64 encoded)

**Comment by Cyan4973** [comment]
Thanks for your greatly detailed investigation.

Level 14 introduces the new compression mode `zstd_btopt` for source files of size <= 256 KB.
This new mode is quite different, and has a lot more work to do.

@Inikep recently identified a slowdown scenario for this mode, which he successfully patched, but only in development version (not part of v0.5.1). 

Not sure though if both issues are really related. My feeling is that the specific pattern of your sample will create a lot of "near matches", and insertion within the binary tree will be slow as a consequence. If that is the case, existing "btlazy2" mode should be impacted too.

Alas, this mode is not present for sources of size <= 128 KB, which seems to be your case. One would need a source size > 128 KB (but still <= 256 KB) to be able to test `btlazy2` at levels 12 et 13.

**Comment by inikep** [comment]
Thanks for report. This problem is fixed in a dev version.
```
zstd v0.5.1 level 13          9.36 MB/s    400 MB/s       110180  47.82 delta_shuffle.txt
zstd v0.5.1 level 14          0.53 MB/s    624 MB/s        99943  43.38 delta_shuffle.txt
```
```
zstd v0.5.2 level 13          9.26 MB/s    368 MB/s       110180  47.82 delta_shuffle.txt
zstd v0.5.2 level 14          3.41 MB/s    649 MB/s        98799  42.88 delta_shuffle.txt
```

**Comment by KrzysFR** [comment]
I used https://github.com/inikep/zstd/commit/6b3739c8e5cc3f8e96d45d72a23708b76fcfa0be and indeed it seems to be fixed. But I'm seeing a bit of regression in compression ratio/time on lower levels.

The delta encoding + shuffled (same as above) seems a bit better:

**Compression time: 0.5.1 vs 0.6.0-dev**
![image](https://cloud.githubusercontent.com/assets/1662235/13222830/05574c8c-d982-11e5-8584-4e034a28a61a.png)
**Compression ratio: 0.5.1 vs 0.6.0-dev**
![image](https://cloud.githubusercontent.com/assets/1662235/13222844/14e63d48-d982-11e5-9b8f-dd0777699547.png)


But the regular encoding + shuffled, though has lost in compression ratio overall from levels 4 to 18:

**Compression time: 0.5.1 vs 0.6.0-dev**
![image](https://cloud.githubusercontent.com/assets/1662235/13222839/0f1fb1fa-d982-11e5-92b6-046ab7e7ddac.png)
**Compression ratio: 0.5.1 vs 0.6.0-dev**
![image](https://cloud.githubusercontent.com/assets/1662235/13222852/1a30f716-d982-11e5-91f1-5e624eb0795b.png)

**Comment by inikep** [comment]
I think a problem is in "zstd_internal.h":
#define ZSTD_OPT_DEBUG 1
which turn optimiaztions for files smaller than 256 KB

Please try zstd with:
#define ZSTD_OPT_DEBUG 0

**Comment by KrzysFR** [comment]
Yes, changing to `ZSTD_OPT_DEBUG ` to 0 fixed it for levels 5 to 18, but now levels 22 to 25 have reverted to the same speed and ratio equivalent to somewhere between 2 and 3 (ie: level 22 is giving me a ratio 2.02 in about 1ms, while level 21 gives 2.32 in 48 ms). I'm guessing the tables for these levels/block size are not yet finalized.

I'll continue keeping an eye on this benchmark to see if this changes with time, but I guess this issue can be considered fixed.

**Comment by inikep** [comment]
The levels 22-25 are just for testing. They are not defined for files smaller than 256 KB so your results can be random :)

**Comment by inikep** [comment]
https://github.com/inikep/zstd/commit/9f754d23dc93c6e2ab60e72d1c0df58bd5a0c5c7

**Comment by KrzysFR** [comment]
Well it's funny because they gave me a tiny bit more compression than level 0, but a little bit faster:
- level 0: ratio **1.70** in **0.81** ms (efficiency 2.09)
- level 22: ratio **1.76** in **0.75** ms (efficiency 2.34)

Another test, I get the same ratio for all levels, but 2x faster than level 0:
- level 0: ratio **4.15** in **0.66** ms (efficiency 6.28)
- level 22: ratio **4.15** in **0.35** ms (efficiency 11.85)

This does not translate to other more regular datasets, but it is interesting that for some, it gives about the same result except faster :)

**Comment by inikep** [comment]
I believe that with old commit (and ZSTD_OPT_DEBUG 0) level 22 was level 0 from the next block of parameters:
{   /* for srcSize <= 256 KB */
    {  0, 18, 19, 18,  0, 12,  4,256, ZSTD_btopt   },  /* level 21*/
{   /* for srcSize <= 128 KB */
    {  0,  0,  0,  0,  0,  0,  0,  0, ZSTD_fast    },  /* level  0 - never used */

So it had only zeros, which was corrected by compressor to default minimum values.

**Comment by Cyan4973** [comment]
Yes, undefined levels ends up with all parameters to 0,
not because it is followed by level 0 in next block,
but because the C standard mandates that non specified values in initialized blocks and structures get set to 0.

Then zstd corrects the params, giving them the minimum possible values.
It ends up being a `ZSTD_fast` strategy with `hashlog == 12`, hence very similar to LZ4.

That it gets faster is not a surprise.
That it gets almost as good or better than level 1 is an interesting surprise.

Regards

**Comment by Cyan4973** [comment]
Should be solved within 0.6.0,
we made some efforts to fix several corner-case scenarios, limiting the slowdown impact.
Hopefully, we haven't found new pathological cases yet.
